Unify error throwing methods

gwbaik9717 opened the pull request:
<!--  Thanks for sending a pull request! -->

#### What this PR does / why we need it?
This PR unifies error throwing methods by refactoring the `logger` to focus solely on logging without throwing errors, and use `YorkieError` to throw errors. 

#### Any background context you want to provide?
The current implementation of the `logger` includes both logging and error throwing, which can lead to confusion and type problems. By separating these concerns, we can ensure that the logger is only responsible for logging messages, while errors are thrown explicitly using `YorkieError`.

#### What are the relevant tickets?
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #696 

### Checklist
- [x] Added relevant tests or not required
- [x] Didn't break anything


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
	- Enhanced error handling across various classes and modules, implementing structured error reporting with `YorkieError` and specific error codes.

- **Bug Fixes**
	- Improved clarity of error messages, ensuring that specific errors are thrown instead of generic ones, aiding in debugging and user feedback.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

gwbaik9717 commented:
@chacha912 

There are some points I'd like to share about the above PR:

> 1. Unify naming convention for Error Code

I have revised the naming convention for error codes. Except for `ok`, I believe it is beneficial for all error code names to start with `Err`.

```typescript
export enum Code {
  // Ok is returned when the operation completed successfully.
  Ok = 'ok',

  // ErrClientNotActivated is returned when the client is not active.
  ErrClientNotActivated = 'ErrClientNotActivated',

  // ErrDummy is returned when the error is intentional.
  ErrDummy = 'ErrDummy',
}
```

> 2. Add New Error Codes `ErrOperationNotReady` and `ErrOperationNotPermitted`

I have added  new Error codes `ErrOperationNotReady`, `ErrOperationNotPermitted`.

`ErrOperationNotReady` is an error when another event needs to be processed first. For example:

```typescript
    fromIdx: number,
    toIdx: number,
    content: string,
    attributes?: A,
  ): [number, number] | undefined {
    if (!this.context || !this.text) {
      const ErrorMessage = 'Text is not initialized yet';
      logger.fatal(ErrorMessage);

      // initialization of `Text` must be preceded ahead of the operation
      throw new YorkieError(Code.ErrOperationNotReady, ErrorMessage);
    }
```

`ErrOperationNotPermitted` is an error when operation is not permitted due to some reasons. For example:

```typescript
  private undo(): void {
    if (this.isUpdating) {
      throw new YorkieError(
        Code.ErrOperationNotPermitted,
        'Undo is not allowed during an update',
      );
    }
```

However, I'm concerned that `ErrOperationNotPermitted` is too general and can be applied to any operations, potentially reducing its specificity. Therefore, I have made an effort to use it as sparingly as possible. 

```typescript
  private undo(): void {
    if (this.isUpdating) {
      throw new YorkieError(
        Code.ErrOperationNotPermitted,
        'Undo is not allowed during an update',
      );
    }
```


> 3. Changes to `YorkieError`'s `toString` method

I've modified the `toString` method of `YorkieError` to remove the `YorkieError:` prefix. 

This change was made because when used with the `logger`, which already includes a `YORKIE [Logger_Level]:` prefix, it seemed redundant.

```typescript
export class YorkieError extends Error {
  name = 'YorkieError';
  stack?: string;

  constructor(readonly code: Code, readonly message: string) {
    super(message);
    this.toString = (): string => `[code=${this.code}]: ${this.message}`;
  }
}
```

```typescript
   if (err instanceof YorkieError) {
        logger.error(err.toString()); // Yorkie E: [code=ErrInvalidArgument]: message
      } else {
        logger.error(err);
      }
```

> 4. Things to be done later...?

I haven't made changes in the `devtools` repository yet because it's a separate repository. I plan to address this in upcoming task (#648) related to monorepo improvements. 

If you have any better suggestions, please let me know.

chacha912 commented on src/document/document.ts:
```diff
@@ -660,7 +660,13 @@ export class Document<T, P extends Indexable = Indexable> {
       this.clone = undefined;
-      logger.error(err);
+
+      if (err instanceof YorkieError) {
+        logger.error(err.toString());
+      } else {
+        logger.error(err);
+      }
+
       throw err;
```
Is it okay if we don't consistently throw `YorkieError` here? 
For example, if we test by intentionally causing an error in the [public/index.html](https://github.com/yorkie-team/yorkie-js-sdk/blob/83a6369f11de43d59d457302283c8ac1c89bdb23/public/index.html#L341) example with `doc.update((root) => {root.a.b = 1;})`, it outputs a general error message.

![image](https://github.com/user-attachments/assets/2fd4bec7-a4b6-46a6-8f1d-4ea32911c662)

gwbaik9717 replied:
Thank you for pointing this out.

In the example you suggested, it seems like there is an intentional error within the `updater`. As expected, the error will be thrown with `throw err` back to the caller in `public/index.html`. When the error is caught in the caller, it is logged using the default `console.error`, which results in the behavior as screenshot above.

```typescript
// document.ts
 try {
      // skipped
      updater(
        proxy,
        new Presence(context, this.clone!.presences.get(actorID)!),
      );
    } catch (err) {
      // skipped
      if (err instanceof YorkieError) {
        logger.error(err.toString());
      } else {
        logger.error(err);
      }

      throw err;
    }
```

```typescript
// public/index.html
async function main(){
   try{
     // intentional error
     doc.update((root) => {root.a.b = 1;})
   }catch(e){
      console.error(e)
   }
}
```

To resolve this, I believe two things need to be done.

> 1.  Replace `console.error` with `logger.error`.

When logging with the default logger such as `console.error`, even if a `YorkieError` is thrown, the error message won't be customized to start with `YORKIE E:`. Therefore, replacing `console.error` with our custom `logger` seems necessary to ensure consistent format of error messages.

> 2. throw `YorkieError` instead of default `err`

Instead of `throw err`, I think it's a great idea to throw a `YorkieError`. For example, it would look like as follows: 

```javascript
// document.ts
 try {
      // skipped
    } catch (err) {
       // skipped
       throw new YorkieError(Code.ErrSomething, "failed to execute updater").
    }
```

However, I'm concerned that it is difficult to determine the exact cause of the error, which makes me hesitant to use `YorkieError`. To be specific, in this case there could be an error in `updater` or just another function within `try` phrase like `createJSON`. Therefore, I doubt that it is a good idea to `throw new YorkieError(Code.ErrSomething, "failed to execute updater").` 

Of course, we could address this issue by using a more general error message, such as `throw new YorkieError(Code.ErrSomething, "failed to update document")`. However, this approach would make it harder to debug the issue in the future.

gwbaik9717 replied:
Moreover, I'm concerned that in the example you showed, I believe there will be two error messages logged with the same content. One logged with `logger.error` and the other with `console.error`. Just to clarify:

```typescript
// document.ts
try{
  updater(
        proxy,
        new Presence(context, this.clone!.presences.get(actorID)!),
      );
}
 catch (err) {
      if (err instanceof YorkieError) {
        logger.error(err.toString());
      } else {
        // logs first time
        logger.error(err);
      }

      throw err
}
```

```typescript
// public/index.html

async function main(){
   try{
     doc.update((root) => {root.a.b = 1;})
   }catch(e){
     // logs second time
      console.error(e)
   }
}
```

Perhaps there is a way to remove one of the logger calls from either the `document` or `public/index.html`, but I'm not sure it's the best way. Please let me know your thoughts on this.

Additionally, I would like to reproduce the error in the snapshot but I'm not sure how to do it. Did you insert the code directly or test it through the console? Any guidance on reproducing the error would be appreciated.

chacha912 replied:
@gwbaik9717 

I added `root.a.b = 1;` to line 345 of `index.html`. 
As you mentioned, this causes the log to be printed twice.

<img width="300" alt="image" src="https://github.com/user-attachments/assets/ce04de43-8ae7-448b-ab6e-364d06554bcb">
<img width="300" alt="image" src="https://github.com/user-attachments/assets/ea88a32d-8d1a-4bb2-a380-c02370bf6ded">

If we set `yorkie.setLogLevel(yorkie.LogLevel.Fatal);` on [line 319](https://github.com/yorkie-team/yorkie-js-sdk/blob/83a6369f11de43d59d457302283c8ac1c89bdb23/public/index.html#L319), the log is printed only once.

<img width="300" alt="image" src="https://github.com/user-attachments/assets/f26240bb-8a0c-4737-90e6-dbcbcf177fda">

Therefore, the visibility of the first log is determined by the logger level set by the user.

chacha912 replied:
As you suggested, I tested wrapping general errors with `YorkieError`, but there doesn't seem to be a significant difference in usability. 

```js
// document.ts
if (err instanceof YorkieError) {
  throw err;
} else {
  throw new YorkieError(Code.ErrDummy, err.toString());
}
```

<img width="1236" alt="image" src="https://github.com/user-attachments/assets/6a3f7086-0da8-4bfa-80eb-4f2eac9f8251">


I initially thought it would be clearer because `YorkieError:` is displayed at the beginning of the error(`[2-1]`), but even when throwing errors normally, we can see from the call stack that the error occurred in `yorkie-js-sdk`.(refer to the blue highlighted section in image `[2-2]`) So it seems like it might be okay not to necessarily wrap errors with YorkieError. Thanks.

( BTW, I discovered that the `message` argument is not being used in the `fatal` method. It's good to remove it. )

<img width="400" alt="image" src="https://github.com/user-attachments/assets/dff6e0ea-ae6a-4061-8599-fc8e4f87c06d">

chacha912 commented on src/document/json/text.ts:
```diff
@@ -285,8 +290,9 @@ export class Text<A extends Indexable = Indexable> {
    */
   toString(): string {
     if (!this.context || !this.text) {
-      logger.fatal('it is not initialized yet');
-      return '';
+      const ErrorMessage = 'Text is not initialized yet';
+      logger.fatal(ErrorMessage);
+      throw new YorkieError(Code.ErrOperationNotReady, ErrorMessage);
     }
```
@hackerwins 
The code seems to have become a bit verbose while separating the logger and throw error logic. There are places where only YorkieError is thrown, and others where the logger is also used.(e.g., it's different in toString and toJSON.) Is there a specific criterion for when to use `Logger.fatal`?

hackerwins replied:
I think the code is becoming verbose with duplicate logging and error throwing. After consideration, it seems better to simplify the approach:

1. Remove the Logger for now: This will reduce code duplication and complexity.
2. Keep only the 'throw new YorkieError' calls: This ensures consistent error handling.
3. Defer logger implementation: We can revisit and implement a comprehensive logging 
   strategy in a future PR.

chacha912 reviewed (commented):
Thank you for your work and detailed explanation.
I'm looking forward to the monorepo setup, which will allow us to work on the code simultaneously in devtools.🙌
I've left a few simple questions; please check the comments.

gwbaik9717 commented on src/util/error.ts:
```diff
@@ -27,8 +27,11 @@ export enum Code {
+  // ErrInvalidType is returned when the type is invalid.
+  ErrInvalidType = 'ErrInvalidType',
+
+  // ErrDummy is returned when the error is intentional.
+  ErrDummy = 'ErrDummy',
```
Thank you for the suggestion. The comment you suggested is much clearer. I'll make a change right away.

gwbaik9717 commented on src/util/error.ts:
```diff
@@ -44,6 +47,12 @@ export enum Code {
   // ErrInvalidArgument is returned when the argument is invalid.
   ErrInvalidArgument = 'ErrInvalidArgument',
+
+  // ErrOperationNotReady is returned when another operation needs to be completed first.
+  ErrOperationNotReady = 'ErrOperationNotReady',
+
+  // ErrOperationNotPermitted is returned when the operation is not permitted.
+  ErrOperationNotPermitted = 'ErrOperationNotPermitted',
```
The document you referred to was very helpful. Based on your suggestions, I have updated the error codes as follows:

```javascript
// ErrNotInitialized is returned when the required initialization has not been completed.
ErrNotInitialized = 'ErrNotInitialized',

// ErrNotReady is returned when the execution of the following actions is not ready.
ErrNotReady = 'ErrNotReady',

// ErrRefused is returned when the execution is rejected.
ErrRefused = 'ErrRefused',
```

Instead of grouping all errors under `ErrNotReady`, I've decided to seperate them to `ErrNotInitialized` and `ErrNotReady` for clarity. Additionally, `ErrNotPermitted` could be misinterpreted as a permission related issue, so I've decided to change error name to `ErrRefused`.

gwbaik9717 commented:
@chacha912 

Thanks to your detailed explanation, I now have a better understanding of different error throwing cases. Based on the suggestions from above, I’ve made the following changes:

Firstly, I removed `logger.fatal` which was being used with `throw new YorkieError` for now due to its verbosity.

Secondly, I removed the `logger.error` part in the `update` method of `Document` in `document.ts` for the same reason: duplicated error logging.

```typescript
// document.ts
try {
     // code skipped
} catch (err) {
      // removed logger.error
      throw err;
}
```

However, unlike `logger.fatal`, which was completely removed for now, I removed `logger.error` only in this specific instance. For example, the following code in `client.ts` uses `logger.error` and throws an error as well, but here, the `logger` provides some useful information that the default logger cannot.

```typescript
// client.ts
.then((res) => {
  // code skipped
})
.catch((err) => {
  logger.error(`[DD] c:"${this.getKey()}" err:`, err);
  this.handleConnectError(err);
  throw err;
});
```

Lastly, I removed the unused `message` parameter from `logger.fatal` as @chacha912  suggested.

Please let me know, if there are any suggestions. :)

chacha912 reviewed (approved):
Thanks for your contribution!
This looks good to me :)
